kong/runloop/balancer.lua

```diff
@@ -526,27 +532,38 @@ do
 
       upstreams_dict[up.name] = up.id
     end
-    return upstreams_dict
+    upstream_version = new_upstream_version
+    all_upstreams = upstreams_dict
+    return true
   end
-  _load_upstreams_dict_into_memory = load_upstreams_dict_into_memory
-
-
-  local opts = { neg_ttl = 10 }
+  _reset_all_upstreams = reset_all_upstreams
 
+  reload_all_upstreams = function()
+    if upstream_version == new_upstream_version then
+      return true
+    end
+    return reset_all_upstreams()
+  end
+end
 
-  ------------------------------------------------------------------------------
-  -- Implements a simple dictionary with all upstream-ids indexed
-  -- by their name.
-  -- @return The upstreams dictionary (a map with upstream names as string keys
-  -- and upstream entity tables as values), or nil+error
-  get_all_upstreams = function()
-    local upstreams_dict, err = singletons.cache:get("balancer:upstreams", opts,
-                                                load_upstreams_dict_into_memory)
-    if err then
-      return nil, err
+local get_all_upstreams
+do
+  local function load_upstreams_dict_into_memory()
+    if kong.configuration.upstream_consistency == "strict" then
+      local ok, err = reload_all_upstreams()
```

Comment:
When `upstream_consistency` is `"strict"` a reload shouldn't be necessary as the upstreams must be already updated. The current behavior, i.e. every upstream change request must be immediately be processed, should be kept in the `"strict"` configuration.